dashboard/test/models/email_preference_test.rb

```diff
@@ -42,15 +42,19 @@ class EmailPreferenceTest < ActiveSupport::TestCase
   end
 
   test "upsert email preference that is already opted in does not opt out" do
-    email_preference = create :email_preference, opt_in: true
+    existing_email_preference = create :email_preference, opt_in: true
+    email = existing_email_preference.email
     EmailPreference.upsert!(
-      email: email_preference.email,
+      email: email,
       opt_in: false,
       ip_address: '172.16.6.1',
       source: EmailPreference::ACCOUNT_SIGN_UP,
       form_kind: 'form text version 3.14159'
     )
-    email_preference.reload
-    assert email_preference.opt_in
+    updated_email_preference = EmailPreference.find_by(email: email)
+    # opt_in is still true.
+    assert updated_email_preference.opt_in
+    # All the attributes of the email preference are the same (no update carried out).
+    assert existing_email_preference == updated_email_preference
```

Comment:
Isn't the `updated_email_preference` the same as the `existing_email_preference`, just a more up-to-date instance? I'd prefer finding by id in that case, or better yet, store away the original attributes and then reload the preference model. Also use `assert_equal` not `assert x == y`. How about something like this:

```ruby
email_preference = create :email_preference, opt_in: true
original_attributes = email_preference.attributes

EmailPreference.upsert!...

assert_equal original_attributes, email_preference.reload.attributes
```

You can of course also explicitly assert `opt_in` after the `reload` if you want to call that out.

Also reflect in the test case name that we expect no attributes to change, not just "does not opt out".

Reply:
I didn't know about the `.attributes` method!  That makes thing simpler.